Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,312 @@
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+
+        foreach (var (id, charges) in component.EntitySpells)
+        {
+            var spell = new EntityTargetAction(_prototypeManager.Index<EntityTargetActionPrototype>(id));
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+    }
+
+    private void OnUse(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        AttemptLearn(uid, component, args);
+
+        args.Handled = true;
+    }
+
+    private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        component.CancelToken = new();
+        var doAfterEventArgs = new DoAfterEventArgs(args.User, component.LearnTime, component.CancelToken.Token, uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            BreakOnDamage = true,
+            BreakOnStun = true,
+            NeedHand = true, //What, are you going to read with your eyes only??
+            TargetFinishedEvent = new LearnDoAfterComplete(args.User),
+            TargetCancelledEvent = new LearnDoAfterCancel(),
+        };
+
+        _doAfter.DoAfter(doAfterEventArgs);
+    }
+
+    private void TeachSpell(EntityUid uid, SpellbookComponent component, LearnDoAfterComplete ev)
+    {
+        component.CancelToken = null;
+
+        if (ev.User == null)
+            return;
+
+        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+    }
+
+    #region Spells
+
+    private void OnRuneMagic(RuneMagicEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+        Spawn(args.RunePrototype, transform.Coordinates.SnapToGrid());
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Teleports the user to the clicked location
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnTeleportSpell(TeleportSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+
+        if (_mapManager.TryFindGridAt(args.Target, out var grid))
+        {
+            var gridPosition = grid.WorldToLocal(args.Target.Position);
+
+            transform.Coordinates = new EntityCoordinates(grid.GridEntityId, gridPosition);
+            SoundSystem.Play(Filter.Pvs(args.Target), args.BlinkSound.GetSound());
+        }
+        else
+        {
+            var mapEntity = _mapManager.GetMapEntityIdOrThrow(args.Target.MapId);
+            transform.AttachParent(mapEntity);
+            transform.WorldPosition = args.Target.Position;
+            SoundSystem.Play(Filter.Pvs(args.Target), args.BlinkSound.GetSound());
+        }
```

Comment:
Can't you just change the mapposition then AttachToGridOrMap?

Also pass in EntityManager to the filters please.

Reply:
Something might be confused here, the args.Target in the sound filter is already map coords.

I might need a better example on how you would want this changed to because I'm having a hard time figuring out. Because what I understand from this is:

-It gets the performers transform component
-It looks for the grid at the world target event target map coords and puts it into a vector 2 for a grid position
- Then it makes new coords for the entity
- Otherwise it looks for the map, then attaches the parent to the new position.
- Plays sound at args.Target (map coords)

I'm not sure how to do it the other way.